### PR TITLE
Okta email authenticator OTP and Magic links integration guide

### PR DESCRIPTION
**Pull request by robertnicolo-okta**
## Description:
Okta email authenticator OTP and Magic links integration guide

### Resolves:

* [OKTA-459279](https://oktainc.atlassian.net/browse/OKTA-459279)

### Deploy link (03/16/2022):

https://62329a7b0760414b9ed96100--reverent-murdock-829d24.netlify.app/docs/guides/authenticators-okta-email/nodeexpress/main/

### CODE REVIEW

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/updateconfigurations.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+1. Click **Save** to save your changes.
+
+### Setup your application to use the email authenticator
+
+#### Add new authentication policy
+
+1. Open the Admin Console for your org.
+1. Choose **Security > Authentication Policies** to show the authentication policies for your org.
+1. Click on **Add a Policy**.
+1. In the **Add Authentication Policy** dialog window, enter a policy name (for example, "2FA Policy) and description.
```
Comment (danielmaharry-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/updateconfigurations.md:20]: I think best practice would be to give the policy a name either matching the app's name as it is a test?

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/overview.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+With the Embedded SDK, your app can verify a user's identity using the email authenticator. You can integrate the email enrollment and challenge in your app and complete those flows using magic links and OTP. Use this guide to learn how.
+
+### Discover the convenience of magic links
+
+A user performs one action with magic links - a click on a link - to complete the email verification. The experience is less choppy than OTP, where a user must switch between their email service and your app to find and submit the OTP. Because of this, magic links is a balance of security and convenience and provides a frictionless experience for your users.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/overview.md:5]: Updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/index.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+<StackSnippet snippet="integrateenrollmagiclink"/> <!-- UNIQUE FOR EACH LANGUAGE  -->
+
+<StackSnippet snippet="integrateenrollotpsummary"/> <!-- UNIQUE FOR EACH LANGUAGE  -->
+
+<StackSnippet snippet="integrateenrollotp" /> <!-- UNIQUE FOR EACH LANGUAGE  -->
+
+## Enable only OTP for the email authenticator
+
+Magic links is a secure way to verify users' emails. However, you may want to use only OTP to provide an even higher level of security that positively proves that the person who started the request is the same person reading the email. In this case, OTP may be a more compatible solution. For example, if you have a banking app, which shows account information and allows for money transfers, magic links may be too convenient, and OTP may provide a better solution.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/index.md:69]: Setting up a meeting with Jeff to discuss.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/index.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+To disable magic link functionality, remove the links from the following email templates:
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/index.md:71]: Updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/index.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+* Email Challenge
+* Email Factor Verification
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/index.md:74]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/designconsiderationscustompasswordrecovery.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
+## Design considerations when customizing magic link for password recovery
+
+### Design Overview
+
+If you've customized your self-service password recovery solution as described in the <StackSnippet snippet="custompwdguide" inline /> and also initiate password recoveries using the [/forgot_password API](/docs/reference/api/users/#forgot-password), you need to consider how your users interact with your magic links. Specifically, your design has the following attributes:
+
+* The magic link `href` attribute in the **Forgot Password** template has been updated to replace the `${resetPasswordLink}` variable with a URL string containing the `otp` and `state` parameters using the `${oneTimePassword}` and `${request.relayState}` variables. An example of the URL string: `http://localhost:8080/login/callback?otp=${oneTimePassword}&state=${request.relayState}`.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/designconsiderationscustompasswordrecovery.md:7]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/designconsiderationscustompasswordrecovery.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+* Your system supports password recoveries using methods other than self-service password recovery. These methods include calling the [/forgot_password API](/docs/reference/api/users/#forgot-password) with the `sendEmail` parameter sent to `true` or enabling password recoveries using the Embedded Sign-in Widget. These methods send an email to the user with a magic link meant to continue the password recovery.
+
+### Considerations
+
+In this design, you've customized the **Forgot Password** email template to include the `${oneTimePassword}` variable per the <StackSnippet snippet="custompwdguide" inline />.  When the [/forgot_password API](/docs/reference/api/users/#forgot-password) and Sign-in Widget are used to initiate the password recovery, an emai is sent to the user. When the user clicks on the magic link and the request is redirected back to your app, the `otp` value is missing. This is because the password recovery was not initiated by the Embedded SDK and does not have context.  Since the `otp` value is missing, your app cannot complete the password reset using the Embedded SDK. Two options are available, however, to complete the password reset with this design. They are:
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/designconsiderationscustompasswordrecovery.md:13]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/getstarted.md [hunk 8]
```diff
@@ -0,0 +1,18 @@
+This guide walks you through the email authenticator user journeys for both magic links and OTP and details step-by-step how to integrate the email enrollment and challenge flows in your app. For more details, see the following sections:
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/getstarted.md:1]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratechallengemagiclink.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        type: "email",
+        key: "okta_email",
+        displayName: "Email",
+      },
+    ],
+  },
+}
+```
+
+Using this response, display a page to input the OTP. Although this use case covers the magic link scenario, displaying an OTP page allows for an OTP verification fallback in cases where OTP may be required or simply more convenient. For example, a user checking their email from a different device must use OTP. [Integrate different browser and device scenario](#integrate-different-browser-and-device-scenario) covers the integration details for the different browser and device scenarios.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratechallengemagiclink.md:45]: Bring this up in meeting with Jeff.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratechallengemagiclink.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+![Magic link in email](/img/authenticators/authenticators-email-challenge-magic-link-in-email.png)
+
+</div>
+
+The link points to your Okta org as in: `https://yourorg.okta.com/email/verify/0oai9ifvveyL3QZ8K696?token=ftr2eAgsg...`
+
+### 5. Send the request to Okta
+
+When the user clicks on the magic link, your org receives the request, gets the `OTP` and `state` parameters, and forwards the request with these parameters to your application. The org uses the `Callback URI` you defined in [Update configuraitons](#update-configurations) to build the URL path and OTP and state query parameters. An example callback URL is as follows: `http://localhost:8080/login/callback?otp=726009&state=1b34371af02dd31d2bc4c48a3607cd32`
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratechallengemagiclink.md:68]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratechallengemagiclink.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+router.get('/login/callback', async (req, res, next) => {
+  const { protocol, originalUrl } = req;
+  const parsedUrl = new URL(protocol + '://' + req.get('host') + originalUrl);
+});
+
+```
+
+### 7. Validate request is from a verification email
+
+Call `OktaAuth.idx.isEmailVerifyCallback()` passing in the query parameter section of the URL. This method validates that the OTP and state query parameters are present. An example of the input parameter is as follows: `?state=8bae50c6e5a973e954b4ac7cd4d1a744&otp=482039`.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratechallengemagiclink.md:84]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratechallengemagiclink.md [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+```javascript
+    if (authClient.idx.isEmailVerifyCallback(search)) {
+      //continue
+    }
+```
+
+### 8. Verify the email and location of the magic link click
+
+Next call `OktaAuth.idx.handleEmailVerifyCallback()` passing in the query parameter containing the OTP and state (for example, `?state=8bae50c6e5a973e954b4ac7cd4d1a744&otp=482039`). This method performs the following checks:
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratechallengemagiclink.md:94]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratechallengemagiclink.md [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+* Ensures that the user clicked the magic link in the same browser (but different tab) as your application.
+* Validates that the `otp` and `state` parameters originate from a nonexpired verification email
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratechallengemagiclink.md:97]: The previous step ensures that the OTP and State are present. This step ensures the browser is the same and the parameters originate from a nonexpired email. Let's talk about this in our next meetup.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratediffbrowserdeviceoverview.md [hunk 14]
```diff
@@ -0,0 +1,13 @@
+## Integrate different browser and device scenario with magic links
+
+Magic links is made with security in mind and only works when there is complete assurance that the person who started the request is the same one who clicked on the magic link. For example, a user who started a sign-in in your app in a web browser must be in the same browser when they click on the magic link. If the browser or device is different, magic links is disabled, and they need to use OTP or be in the same browser to complete the email verification. The following flowchart illustrates this logic.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratediffbrowserdeviceoverview.md:3]: Updated and also will bring it up in a Jeff meeting.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratediffbrowserdeviceoverview.md [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+
+<div class="common-image-format">
+
+![Diagram showing nagic link flow for same and diff browsers](/img/authenticators/authenticators-email-magic-link-flowchart.png)
+
+</div>
+
+The following step-by-step instructions detail integrating the different browser scenario using the email challenge.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratediffbrowserdeviceoverview.md:11]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratediffbrowserdevice.md [hunk 16]
```diff
@@ -0,0 +1,32 @@
+### 1. Initiate sign and start magic link flow
+
+First, the user initiates the sign-in and email challenge flow. Next, they click on the magic link, which redirects them back to your app. Your app validates that the `otp` and `state` parameters are included in the URL. The steps are identical to the first steps in [Integrate email challenge with magic links](#integrate-email-challenge-with-magic-links).
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratediffbrowserdevice.md:3]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratediffbrowserdevice.md [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+
+### 2. Verify the email and location of magic link click
+
+Next call `OktaAuth.idx.handleEmailVerifyCallback()` passing in the query parameter containing the OTP and state (for example, `?state=8bae50c6e5a973e954b4ac7cd4d1a744&otp=482039`). This method performs the following checks:
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratediffbrowserdevice.md:8]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratediffbrowserdevice.md [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+
+#### 3: Handle error and inform user to use OTP on original tab
+
+`OktaAuth.idx.handleEmailVerifyCallback()` throws a `EmailVerifyCallbackError` error if the user clicked on the magic link in a different browser or device. Handle the error and inform the user to enter the OTP on the original tab.
+
+<div class="common-image-format">
+
+![Screenshot of different device or browser error](/img/authenticators/authenticators-email-magic-link-error.png)
+
+</div>
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integratediffbrowserdevice.md:32]: Will bring it up with Jeff.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integrateenrollmagiclink.md [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+First, the user initiates the sign-in with username and password by making a call to `OktaAuth.idx.authenticate()`.
+
+```javascript
+  const authClient = getAuthClient(req);
+  const transaction = await authClient.idx.authenticate({ username,password});
+```
+
+### 2. Show option to enroll email
+
+If email enrollment is enabled for you org's application and the user has not yet enrolled in email, `OktaAuth.idx.authenticate()` returns `IdxTransaction` with a status of `PENDING` and `nextStep.name` equal to `select-authenticator-enroll`. The `nextStep.options` array includes an item with a `value` property of `okta_email`.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integrateenrollmagiclink.md:13]: Updated, but this markdown is commented out. Enrollment not yet supported for backend.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integrateenrollmagiclink.md [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+      type: "email",
+      authenticator: {
+        key: "okta_email",
+      },
+    },
+  ],
+}
+```
+
+Using this response display a page to input the OTP. Although this use case covers the magic link scenario, displaying an OTP page allows for a OTP verification fallback in cases where OTP maybe required or simply more convienent.  For example, a user checking their email from a different device from where they are signing into your app, is required to use OTP. Different browser and device scenarios are covered in  [Integrate different browser and device scenario][#integrate-different-browser-and-device-scenario]
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integrateenrollmagiclink.md:83]: Updated, but this markdown is commented out. Enrollment not yet supported for backend.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integrateenrollmagiclink.md [hunk 21]
```diff
@@ -0,0 +1,168 @@
+<!-- WEB SHARED: KEEP EMPTY FOR JAVA, JAVASCRIPT, .NET, AND GO
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integrateenrollmagiclink.md:1]: This markdown is commented out. You can leave it blank for .NET.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integrateenrollmagiclink.md [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+
+Next, the user opens their email and clicks on the magic link. The following screenshots shows the magic link in the email.
+
+<div class="common-image-format">
+
+![Magic link in email](/img/authenticators/authenticators-email-enroll-magic-link-in-email.png)
+
+</div>
+
+The link is points to your Okta org as in: `https://yourorg.okta.com/email/verify/0oai9ifvveyL3QZ8K696?token=ftr2eAgsg...`
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integrateenrollmagiclink.md:101]: Updated, but this md is commented out.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integrateenrollmagiclink.md [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+
+### 7. Send request to Okta
+
+When the user clicks on the magic link, the request is first routed to you Okta org, which forwards the request to your application including the `OTP` and `state` parameters. The org uses the `Callback URI` you defined in [Update configuraitons](#update-configurations) to build the URL path and adds the OTP and state query parameters. An example callback URL: `http://localhost:8080/login/callback?otp=726009&state=1b34371af02dd31d2bc4c48a3607cd32`
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integrateenrollmagiclink.md:105]: Updated though this markdown is commented out.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integrateenrollmagiclink.md [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+router.get('/login/callback', async (req, res, next) => {
+  const { protocol, originalUrl } = req;
+  const parsedUrl = new URL(protocol + '://' + req.get('host') + originalUrl);
+});
+
+```
+
+### 9. Validate request is from a verification email
+
+Call `OktaAuth.idx.isEmailVerifyCallback()` passing in the query parmaeter section of the URL. This method validates that the OTP and state query parameters are present. `?state=8bae50c6e5a973e954b4ac7cd4d1a744&otp=482039` is an example of the method's input parmaeter.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/integrateenrollmagiclink.md:121]: commented out.

File: packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/updateconfigurations.md (hunk 1 above)
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-okta-email/main/nodeexpress/updateconfigurations.md:20]: For consistency, I'll defer to you on the actual implementation of these instructions.

**Pull request merged**